### ISSUE TITLE
Allow IHostBuilder to have the convenient UseStartup<> pattern that IWebHostBuilder implements

### ISSUE THREAD

**Issue by sonicmouse** [dotnet/runtime#42258]
### Preamble
I would like to see a way to use the convenient `Startup.cs` pattern that `IWebHostBuilder` has implemented, but at the `IHostBuilder` level.

I don't like working in the `Program.cs` file... That is typically for boiler-plate code to start a host. Now I find myself in there constantly making changes just to work on my DI pipeline, for example:

**Program.cs:**

	// ...
	public static IHostBuilder CreateHostBuilder(string[] args) =>
		Host.CreateDefaultBuilder(args)
		.ConfigureServices(services =>
		{
			// add services here
		});
	// ...

### I'd like to see...
It would be really convenient to segregate the boiler-plate code from the DI injection routines. So instead of the code above, do this:

	// ...
	public static IHostBuilder CreateHostBuilder(string[] args) =>
		Host.CreateDefaultBuilder(args)
		.UseStartup<Startup>(); // new method
	// ...

### Additional context
I put together an extension that makes this possible, but I am not 100% sure this is proper. Maybe there is a better way you folks can come up with that would accomplish this in a more organic manner:

	/// <summary>
	/// Extensions to emulate a typical "Startup.cs" pattern for <see cref="IHostBuilder"/>
	/// </summary>
	public static class HostBuilderExtensions
	{
		private const string ConfigureServicesMethodName = "ConfigureServices";

		/// <summary>
		/// Specify the startup type to be used by the host.
		/// </summary>
		/// <typeparam name="TStartup">The type containing an optional constructor with
		/// an <see cref="IConfiguration"/> parameter. The implementation should contain a public
		/// method named ConfigureServices with <see cref="IServiceCollection"/> parameter.</typeparam>
		/// <param name="hostBuilder">The <see cref="IHostBuilder"/> to initialize with TStartup.</param>
		/// <returns>The same instance of the <see cref="IHostBuilder"/> for chaining.</returns>
		public static IHostBuilder UseStartup<TStartup>(
			this IHostBuilder hostBuilder) where TStartup : class
		{
			// Invoke the ConfigureServices method on IHostBuilder...
			hostBuilder.ConfigureServices((ctx, serviceCollection) =>
			{
				// Find a method that has this signature: ConfigureServices(IServiceCollection)
				var cfgServicesMethod = typeof(TStartup).GetMethod(ConfigureServicesMethodName,
					new Type[] { typeof(IServiceCollection) });

				// Check if TStartup has a ctor that takes a IConfiguration parameter
				var hasConfigCtor = typeof(TStartup).GetConstructor(
					new Type[] { typeof(IConfiguration) }) != null;

				// create a TStartup instance based on ctor
				var startUpObj = hasConfigCtor ?
					(TStartup)Activator.CreateInstance(typeof(TStartup), ctx.Configuration) :
					(TStartup)Activator.CreateInstance(typeof(TStartup), null);

				// finally, call the ConfigureServices implemented by the TStartup object
				cfgServicesMethod?.Invoke(startUpObj, new object[] { serviceCollection });
			});

			// chain the response
			return hostBuilder;
		}
	}

Thank you for taking the time to look at this request.

**Comment by eerhardt** [comment]
Since this is a proposal for adding API, is someone willing make the proposal in the format of the template linked in step 1 of https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md ?

**Comment by EccentricVamp** [comment]
Thanks @Tratcher  for the thoughtful response and proposal. This solution seems adequate to me

@eerhardt  I gave the template a shot #42364
I feel it is necessary to point out that I'm not the one who originally opened this request, and was merely looking for an elegant/standard solution for configuring services. So, I mostly just copied suggestions from this thread. Let me know if I messed anything up, or missed something
